codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/RuleSetCreationSpec.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.rules;
+
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.TreeNode;
+import com.fasterxml.jackson.jr.stree.JrsArray;
+import com.fasterxml.jackson.jr.stree.JrsBoolean;
+import com.fasterxml.jackson.jr.stree.JrsNumber;
+import com.fasterxml.jackson.jr.stree.JrsObject;
+import com.fasterxml.jackson.jr.stree.JrsString;
+import com.fasterxml.jackson.jr.stree.JrsValue;
+import com.squareup.javapoet.CodeBlock;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import software.amazon.awssdk.codegen.model.intermediate.IntermediateModel;
+import software.amazon.awssdk.codegen.model.rules.endpoints.ConditionModel;
+import software.amazon.awssdk.codegen.model.rules.endpoints.EndpointModel;
+import software.amazon.awssdk.codegen.model.rules.endpoints.ParameterDeprecatedModel;
+import software.amazon.awssdk.codegen.model.rules.endpoints.ParameterModel;
+import software.amazon.awssdk.codegen.model.rules.endpoints.RuleModel;
+import software.amazon.awssdk.codegen.model.service.EndpointRuleSetModel;
+import software.amazon.awssdk.core.rules.Condition;
+import software.amazon.awssdk.core.rules.Endpoint;
+import software.amazon.awssdk.core.rules.EndpointRuleset;
+import software.amazon.awssdk.core.rules.Expr;
+import software.amazon.awssdk.core.rules.FnNode;
+import software.amazon.awssdk.core.rules.Identifier;
+import software.amazon.awssdk.core.rules.Parameter;
+import software.amazon.awssdk.core.rules.ParameterType;
+import software.amazon.awssdk.core.rules.Parameters;
+import software.amazon.awssdk.core.rules.Rule;
+
+public class RuleSetCreationSpec {
```

Comment:
Nice work making this class so easy to read.

Reply:
thanks! I spent quite a bit of time thinking about how to organize this